### DISCUSSION TITLE
Making flatmaps with pycortex

### DISCUSSION THREAD

**Discussion by uhussai7** [kendrickkay/nsdcode#23] [Q&A] [answered]
Just trying to figure out how I would go about making flat maps in pycortex with the NSD data. My understanding so far is the following:

- [`pycortex`](https://github.com/gallantlab/pycortex): Here I carry out the following commands:
    - `cortex.freesurfer.import_subj("subj01",freesurfer_subject_dir=freesurfer_path)`, where `freesurfer_path` is the freesurfer folder for each subject in the nsd dataset.
    - `cortex.freesurfer.import_flat("subj01")`
    - I try to follow [this example](http://gallantlab.org/pycortex/auto_examples/datasets/plot_volume_to_vertex.html#sphx-glr-auto-examples-datasets-plot-volume-to-vertex-py) after importing but `cortex.db.subj01.transforms` is empty. Should it have been populated during the import?
    - Further the `overlays.svg` file does not contain the ROIs and their names, how can one get those?
- [`nsd_mapdata.py`](https://github.com/cvnlab/nsdcode/blob/master/nsdcode/nsd_mapdata.py): My guess is that this function is needed to map from functional space (1.8mm betas) to surfaces. Although not clear how I can then use the output of this with `pycortex`.

Been trying to find a way to resolve this and it's been somewhat time consuming. Any guidance would be greatly appreciated!

**Comment by kendrickkay** [comment]
I am not deeply familiar with pycortex, and it may be the case that pycortex and NSD have somewhat differing data/api formats. Perhaps this question is also a question for the pycortex developers. I will try to ping some people that might have some ideas.

**Reply by uhussai7** [reply]
Thank you for your prompt response! FYI: Any other package/approach is also totally fine, just need to show annotated ROIs with fmri on a flat map.

**Reply by AbabababaSmart** [reply]
Hi, could you please share the pycortex db and the svg with me (^人^)

**Comment by ariaaay** [comment, marked as answer]
In terms of the svg overlay, we have to hand draw ROIs in Inkscape with the functional localizer data, which will get saved to the SVG file. I could share the pycortex db and the svg with you if you would like. See https://github.com/ariaaay/clip2brain/wiki/General-instructions-for-creating-Pycortex-Flatmaps for general instructions.

**Reply by uhussai7** [reply]
Thank you for your instructions. The pycortex db would be super helpful thanks!

**Reply by ariaaay** [reply]
https://drive.google.com/drive/folders/1YXAYggtlsfFxMtiqWs0UrMihxI125MHb?usp=share_link

**Reply by ariaaay** [reply]
Will update the svg in a bit.

**Reply by uhussai7** [reply]
Beautiful thanks so much!

**Reply by ariaaay** [reply]
You are welcomed!
SVGs are updated. Also you can adjust which ROI to show in Inkscape. Showing all ROIs will crowd the space too much since ROIs overlap.

**Reply by uhussai7** [reply]
Just a follow up question if you get a chance: How were the transforms in the folders /subjXX/transforms/func1pt8_to_anat0pt8_autoFSbbr/ calculated? My understanding is that nsd_mapdata handles the mapping but how does one extract the transform?

**Reply by kendrickkay** [reply]
Just to chime in. It sounds like the pycortex preparation may have involved its own determination of alignment (e.g. func 1.8mm to anat 0.8mm).  In general, since the functional data were acquired on a 7T scanner and the anatomical data were acquired on a different 3T scanner, the geometries of data coming from the two scanners were different, and necessitated a moderately nonlinear warp.  nsd_mapdata takes care of this.  Not sure how pycortex was prepared.

**Reply by uhussai7** [reply]
I ended up just writing a bit of code to make the flat maps on mayavi, using the nsd_mapdata function to go into anat space ofcourse. Thanks!

**Reply by Boltzmachine** [reply]
Hi just wonder where I should put the db folder to let pycortex find it

**Reply by AbabababaSmart** [reply]
Hi, could you please share the pycortex db and the svg with me (^人^)

**Comment by maggiemae13** [comment]
It's been a while since I've used pycortex, but from what I remember:

- Importing a subject does not import any transforms automatically. In order to project data onto a surface using cortex.Volume, there are a couple options:
    - Use the pycortex align function to align functional space data to the anatomical reference imported from freesurfer.
    - Import an appropriate transform from FSL or Freesurfer into pycortex using cortex.xfm.Transform.from_XXXX. 
    - Use nsd_mapdata to transform data in functional space into anatomical space. When viewing it in pycortex you can use the built-in 'identity' transform. Note, that sometimes coordinate systems can get messed up when moving between different software. After loading functional data that is already in anatomical space you may need to do something like this to get it in the correct orientation: np.flip((np.rot90(DATA_ARRAY, k=3)),2)

The ROI overlays you also have to create/add yourself. If you have freesurfer annotations this thread may be useful:
https://github.com/gallantlab/pycortex/issues/312

**Reply by uhussai7** [reply]
Thank you for your response. This is helpful!

**Comment by geoffreyfan** [comment]
Hello, could you please share the pycortex db file with me? Thank you very much!
